OpenSSL 3.5.0: undeclared function 'SSL_SESSION_get_time'

### Environment

Include the result of the following commands:
  - FreeBSD build 13.5-RELEASE FreeBSD 13.5-RELEASE CORE amd64

### Description

Compiling Nginx 1.26.3 on FreeBSD "--with-openssl=/usr/src/openssl-3.5.0" fails,

cc -c -pipe  -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Wno-unused-parameter -Werror -g -I/usr/local/include -I src/core  -I src/event  -I src/event/modules  -I src/event/quic  -I src/os/unix  -I /usr/src/pcre2-10.45/src/  -I /usr/src/openssl-3.5.0/.openssl/include  -I /usr/src/zlib-1.3.1  -I objs  -o objs/src/event/ngx_event_openssl.o  src/event/ngx_event_openssl.c
src/event/ngx_event_openssl.c:1153:20: error: call to undeclared function 'SSL_SESSION_get_time'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
 1153 |             time = SSL_SESSION_get_time(sess);
      |                    ^
src/event/ngx_event_openssl.c:1163:17: error: call to undeclared function 'SSL_SESSION_set_time'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
 1163 |                 SSL_SESSION_set_time(sess, now);
      |                 ^
2 errors generated.
*** Error code 1

Looks like `long SSL_SESSION_get_time(const SSL_SESSION *s);` and `long SSL_SESSION_set_time(SSL_SESSION *s, long tm);` are deprecated in OpenSSL 3.4
https://docs.openssl.org/3.4/man3/SSL_SESSION_get_time/#name

Thanks. I removed `no-deprecated` from the openssl-3.5.0 build, though it looks like `no-dh` has tripped up the build,
```
cc -c -pipe  -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Wno-unused-parameter -Werror -g -I/usr/local/include -I src/core  -I src/event  -I src/event/modules  -I src/event/quic  -I src/os/unix  -I ../pcre2-10.45/src/  -I ../openssl-3.5.0/.openssl/include  -I ../zlib-1.3.1  -I objs  -o objs/src/event/ngx_event_openssl.o  src/event/ngx_event_openssl.c
src/event/ngx_event_openssl.c:1431:10: error: call to undeclared function 'PEM_read_bio_DHparams'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
 1431 |     dh = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
      |          ^
src/event/ngx_event_openssl.c:1431:10: note: did you mean 'PEM_read_bio_Parameters'?
../openssl-3.5.0/.openssl/include/openssl/pem.h:526:11: note: 'PEM_read_bio_Parameters' declared here
  526 | EVP_PKEY *PEM_read_bio_Parameters(BIO *bp, EVP_PKEY **x);
      |           ^
src/event/ngx_event_openssl.c:1431:8: error: incompatible integer to pointer conversion assigning to 'DH *' (aka 'struct dh_st *') from 'int' [-Wint-conversion]
 1431 |     dh = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
      |        ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/event/ngx_event_openssl.c:1442:9: error: call to undeclared function 'DH_free'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
 1442 |         DH_free(dh);
      |         ^
src/event/ngx_event_openssl.c:1442:9: note: did you mean 'BN_free'?
../openssl-3.5.0/.openssl/include/openssl/bn.h:305:6: note: 'BN_free' declared here
  305 | void BN_free(BIGNUM *a);
      |      ^
src/event/ngx_event_openssl.c:1447:5: error: call to undeclared function 'DH_free'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
 1447 |     DH_free(dh);
      |     ^
4 errors generated.
*** Error code 1
```

--- a/src/event/ngx_event_openssl.c
+++ b/src/event/ngx_event_openssl.c
@@ -1315,6 +1315,8 @@ ngx_ssl_passwords_cleanup(void *data)
 ngx_int_t
 ngx_ssl_dhparam(ngx_conf_t *cf, ngx_ssl_t *ssl, ngx_str_t *file)
 {
+#ifndef OPENSSL_NO_DH
+
     BIO  *bio;
 
     if (file->len == 0) {
@@ -1384,6 +1386,8 @@ ngx_ssl_dhparam(ngx_conf_t *cf, ngx_ssl_t *ssl, ngx_str_t *file)
 #endif
 
     BIO_free(bio);
+
+#endif
 
     return NGX_OK;
 }

--- a/src/event/ngx_event_openssl.h
+++ b/src/event/ngx_event_openssl.h
@@ -19,7 +19,9 @@
 #include <openssl/bn.h>
 #include <openssl/conf.h>
 #include <openssl/crypto.h>
+#ifndef OPENSSL_NO_DH
 #include <openssl/dh.h>
+#endif
 #ifndef OPENSSL_NO_ENGINE
 #include <openssl/engine.h>
 #endif

--- a/src/event/ngx_event_openssl.h
+++ b/src/event/ngx_event_openssl.h
@@ -83,6 +83,17 @@
 #endif
 
 
+#ifdef OPENSSL_NO_DEPRECATED_3_4
+#define SSL_SESSION_get_time(s)      SSL_SESSION_get_time_ex(s)
+#define SSL_SESSION_set_time(s, t)   SSL_SESSION_set_time_ex(s, t)
+#endif
+
+
+#ifdef OPENSSL_NO_DEPRECATED_3_0
+#define EVP_CIPHER_CTX_cipher(c)     EVP_CIPHER_CTX_get0_cipher(c)
+#endif
+
+
 typedef struct ngx_ssl_ocsp_s   ngx_ssl_ocsp_t;
 
 

